File: lib/change/root-resolution.js
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";

export function resolveChangeContext(input = {}) {
  const cwd = path.resolve(input.cwd ?? process.cwd());
  const env = input.env ?? process.env;
  const cwdChange = findCwdChange(cwd);
  const changeId = input.changeId ?? input.change_id ?? cwdChange?.changeId ?? null;
  const stateRoot = resolveExplicitRoot(input.stateRoot ?? input.state_root, cwd, "state-root");
  const repoRoot = resolveExplicitRoot(input.repoRoot ?? input.repo_root, cwd, "repo-root");
  const codeRoot = resolveCodeRoot(input.codeRoot ?? input.code_root, cwd);

  const aliasedRoot = [stateRoot, repoRoot].find((root) => root && root.requested !== root.resolved);
  if (aliasedRoot) {
    return unresolvedContext({
      codeRoot,
      changeId,
      source: "unresolved",
      unresolvedReason: "aliased-state-root",
      candidates: [explicitCandidate(aliasedRoot.resolved, changeId, aliasedRoot.source, aliasedRoot.requested)]
    });
  }

  if (stateRoot && repoRoot && !samePath(stateRoot.resolved, repoRoot.resolved)) {
    return unresolvedContext({
      codeRoot,
      changeId,
      source: "conflict",
      unresolvedReason: "conflicting-explicit-roots",
      candidates: [
        explicitCandidate(stateRoot.resolved, changeId, "state-root", stateRoot.requested),
        explicitCandidate(repoRoot.resolved, changeId, "repo-root", repoRoot.requested)
      ]
    });
  }

  if (stateRoot || repoRoot) {
    const selected = stateRoot ?? repoRoot;
    if (isLinkedWorktree(selected.resolved, gitWorktrees(selected.resolved))) {
      return unresolvedContext({
        codeRoot,
        changeId,
        source: "unresolved",
        unresolvedReason: "linked-worktree-state-root",
        isLinked: true,
        candidates: [explicitCandidate(selected.resolved, changeId, selected.source, selected.requested)]
      });
    }
    return resolvedContext({
      stateRoot: selected.resolved,
      codeRoot,
      changeId,
      source: selected.source,
      candidates: [explicitCandidate(selected.resolved, changeId, selected.source, selected.requested)]
    });
  }

  const envRoot = resolveExplicitRoot(env.HARNESS_CHANGE_STATE_ROOT, cwd, "environment");
  if (envRoot) {
    if (envRoot.requested !== envRoot.resolved) {
      return unresolvedContext({
        codeRoot,
        changeId,
        source: "unresolved",
        unresolvedReason: "aliased-state-root",
        candidates: [explicitCandidate(envRoot.resolved, changeId, "environment", envRoot.requested)]
      });
    }
    if (isLinkedWorktree(envRoot.resolved, gitWorktrees(envRoot.resolved))) {
      return unresolvedContext({
        codeRoot,
        changeId,
        source: "unresolved",
        unresolvedReason: "linked-worktree-state-root",
        isLinked: true,
        candidates: [explicitCandidate(envRoot.resolved, changeId, "environment", envRoot.requested)]
      });
    }
    return resolvedContext({
      stateRoot: envRoot.resolved,
      codeRoot,
      changeId,
      source: "environment",
      candidates: [explicitCandidate(envRoot.resolved, changeId, "environment", envRoot.requested)]
    });
  }

  const worktrees = gitWorktrees(codeRoot);
  const isLinked = isLinkedWorktree(codeRoot, worktrees);
  const candidates = [];
  const knownWorktreeCandidates = [];

  if (changeId && cwdChange?.changeId === changeId) {
    addCandidate(candidates, cwdChange.stateRoot, changeId, "cwd-change", true);
  }
  if (changeId) {
    addCandidate(candidates, codeRoot, changeId, "code-root", true);
  }
  for (const worktreeRoot of worktrees) {
    const candidate = explicitCandidate(worktreeRoot, changeId, "git-worktree");
    knownWorktreeCandidates.push(candidate);
    if (changeId && candidate.exists) {
      addCandidate(candidates, worktreeRoot, changeId, samePath(worktreeRoot, codeRoot) ? "code-root" : "git-worktree", true);
    }
  }

  if (candidates.length > 1) {
    return unresolvedContext({
      codeRoot,
      changeId,
      source: "unresolved",
      unresolvedReason: "ambiguous-state-root",
      isLinked,
      candidates
    });
  }

  if (candidates.length === 1) {
    const candidate = candidates[0];
    if (!samePath(candidate.state_root, codeRoot) && candidate.source === "git-worktree") {
      return unresolvedContext({
        codeRoot,
        changeId,
        source: "unresolved",
        unresolvedReason: "other-worktree-state-root",
        isLinked,
        candidates
      });
    }
    return resolvedContext({
      stateRoot: candidate.state_root,
      codeRoot,
      changeId,
      source: candidate.source,
      isLinked,
      candidates
    });
  }

  if (isLinked) {
    return unresolvedContext({
      codeRoot,
      changeId,
      source: "unresolved",
      unresolvedReason: "linked-worktree-unresolved",
      isLinked,
      candidates: knownWorktreeCandidates
    });
  }

  return resolvedContext({
    stateRoot: codeRoot,
    codeRoot,
    changeId,
    source: cwdChange ? "cwd-change" : "cwd",
    isLinked,
    candidates
  });
}

export function resolveChangeContextFromArgs(args, options = {}) {
  return resolveChangeContext({
    stateRoot: args.values.get("state-root"),
    repoRoot: args.values.get("repo-root"),
    codeRoot: args.values.get("code-root"),
    changeId: options.changeId,
    cwd: options.cwd,
    env: options.env
  });
}

export function writeRootContextError(context, stream = process.stderr) {
  stream.write(`${rootContextErrorText(context)}\n`);
}

export function rootContextErrorText(context) {
  const change = context.change_id ? ` for change ${context.change_id}` : "";
  if (context.unresolved_reason === "conflicting-explicit-roots") {
    const state = context.candidates.find((candidate) => candidate.source === "state-root")?.state_root ?? "<missing>";
    const repo = context.candidates.find((candidate) => candidate.source === "repo-root")?.state_root ?? "<missing>";
    return [
      `ERROR: conflicting state roots${change}: --state-root ${state} differs from --repo-root ${repo}`,
      "GUIDE: retry with one canonical state root."
    ].join("\n");
  }
  if (context.unresolved_reason === "aliased-state-root") {
    const candidate = context.candidates[0];
    return [
      `ERROR: state root alias${change}: ${candidate?.requested_root ?? "<missing>"} resolves to ${candidate?.state_root ?? "<missing>"}`,
      "GUIDE: retry with the canonical state-root path."
    ].join("\n");
  }
  if (context.unresolved_reason === "linked-worktree-state-root") {
    const candidate = context.candidates[0];
    return [
      `ERROR: state root${change} is a linked worktree: ${candidate?.state_root ?? "<missing>"}`,
      "GUIDE: retry with the canonical shared state-root, not a linked worktree."
    ].join("\n");
  }
  if (context.unresolved_reason === "ambiguous-state-root") {
    return [
      `ERROR: ambiguous state root${change}`,
      ...candidateLines(context),
      "GUIDE: inspect candidates and retry with --state-root <canonical-state-root>."
    ].join("\n");
  }
  if (context.unresolved_reason === "other-worktree-state-root") {
    return [
      `ERROR: state root unresolved${change}; matching change workspace is in another worktree`,
      ...candidateLines(context),
      "GUIDE: retry with explicit --state-root after confirming the canonical workspace."
    ].join("\n");
  }
  return [
    `ERROR: state root unresolved${change}`,
    ...candidateLines(context),
    "GUIDE: retry with --state-root <canonical-state-root>."
  ].join("\n");
}

function resolvedContext({ stateRoot, codeRoot, changeId, source, isLinked = false, candidates = [] }) {
  return {
    state_root: stateRoot,
    code_root: codeRoot,
    change_id: changeId,
    source,
    is_linked_worktree: isLinked,
    candidates,
    unresolved_reason: null
  };
}

function unresolvedContext({ codeRoot, changeId, source, unresolvedReason, isLinked = false, candidates = [] }) {
  return {
    state_root: null,
    code_root: codeRoot,
    change_id: changeId,
    source,
    is_linked_worktree: isLinked,
    candidates,
    unresolved_reason: unresolvedReason
  };
}

function candidateLines(context) {
  return context.candidates.map(
    (candidate) =>
      `CANDIDATE: ${candidate.state_root} source=${candidate.source} exists=${candidate.exists ? "yes" : "no"}`
  );
}

function explicitCandidate(stateRoot, changeId, source, requestedRoot = null) {
  return {
    state_root: stateRoot,
    change_dir: changeId ? path.join(stateRoot, ".changes", changeId) : null,
    source,
    exists: changeId ? fs.existsSync(path.join(stateRoot, ".changes", changeId)) : fs.existsSync(stateRoot),
    ...(requestedRoot ? { requested_root: requestedRoot } : {})
  };
}

function addCandidate(candidates, stateRoot, changeId, source, requireExists) {
  const candidate = explicitCandidate(stateRoot, changeId, source);
  if (requireExists && !candidate.exists) {
    return;
  }
  if (candidates.some((entry) => samePath(entry.state_root, candidate.state_root))) {
    return;
  }
  candidates.push(candidate);
}

function resolveRoot(value, cwd) {
  return resolveExplicitRoot(value, cwd, "explicit")?.resolved ?? null;
}

function resolveExplicitRoot(value, cwd, source) {
  if (!value) {
    return null;
  }
  const requested = path.resolve(cwd, value);
  return { requested, resolved: realPathOrResolve(requested), source };
}

function resolveCodeRoot(value, cwd) {
  const explicit = resolveRoot(value, cwd);
  if (explicit) {
    return explicit;
  }
  return realPathOrResolve(gitRoot(cwd) ?? cwd);
}

function gitRoot(cwd) {
  const result = spawnSync("git", ["-C", cwd, "rev-parse", "--show-toplevel"], { encoding: "utf8" });
  if (result.status !== 0) {
    return null;
  }
  return path.resolve(result.stdout.trim());
}

function gitWorktrees(codeRoot) {
  const result = spawnSync("git", ["-C", codeRoot, "worktree", "list", "--porcelain"], { encoding: "utf8" });
  if (result.status !== 0) {
    return [];
  }
  return result.stdout
    .split("\n")
    .filter((line) => line.startsWith("worktree "))
    .map((line) => path.resolve(line.slice("worktree ".length).trim()))
    .filter(Boolean);
}

function isLinkedWorktree(codeRoot, worktrees) {
  if (worktrees.length <= 1) {
    return false;
  }
  const current = realPathOrResolve(codeRoot);
  return worktrees.some((worktreeRoot, index) => index > 0 && samePath(realPathOrResolve(worktreeRoot), current));
}

function findCwdChange(cwd) {
  const parts = path.resolve(cwd).split(path.sep);
  for (let index = parts.length - 2; index >= 0; index -= 1) {
    if (parts[index] !== ".changes" || !parts[index + 1]) {
      continue;
    }
    const stateRootParts = parts.slice(0, index);
    const stateRoot = stateRootParts.length === 0 ? path.sep : stateRootParts.join(path.sep) || path.sep;
    return {
      stateRoot: path.resolve(stateRoot),
      changeId: parts[index + 1],
      changeDir: path.join(path.resolve(stateRoot), ".changes", parts[index + 1])
    };
  }
  return null;
}

function samePath(left, right) {
  return realPathOrResolve(left) === realPathOrResolve(right);
}

function realPathOrResolve(value) {
  try {
    return fs.realpathSync(value);
  } catch {
    return path.resolve(value);
  }
}

File: lib/change/root_resolution.py
#!/usr/bin/env python3
from __future__ import annotations

import os
import subprocess
from pathlib import Path


def resolve_change_context(
    *,
    state_root: str | None = None,
    repo_root: str | None = None,
    code_root: str | None = None,
    change_id: str | None = None,
    cwd: str | Path | None = None,
    env: dict | None = None,
) -> dict:
    cwd_path = Path(cwd or os.getcwd()).resolve()
    env_map = env if env is not None else os.environ
    cwd_change = find_cwd_change(cwd_path)
    resolved_change = change_id or (cwd_change or {}).get("change_id")
    explicit_state = resolve_explicit_root(state_root, cwd_path, "state-root")
    legacy_repo = resolve_explicit_root(repo_root, cwd_path, "repo-root")
    resolved_code_root = resolve_code_root(code_root, cwd_path)

    aliased_root = next(
        (root for root in (explicit_state, legacy_repo) if root and root["requested"] != root["resolved"]),
        None,
    )
    if aliased_root:
        return unresolved_context(
            code_root=resolved_code_root,
            change_id=resolved_change,
            source="unresolved",
            unresolved_reason="aliased-state-root",
            candidates=[candidate(aliased_root["resolved"], resolved_change, aliased_root["source"], aliased_root["requested"])],
        )

    if explicit_state and legacy_repo and explicit_state["resolved"] != legacy_repo["resolved"]:
        return unresolved_context(
            code_root=resolved_code_root,
            change_id=resolved_change,
            source="conflict",
            unresolved_reason="conflicting-explicit-roots",
            candidates=[
                candidate(explicit_state["resolved"], resolved_change, "state-root", explicit_state["requested"]),
                candidate(legacy_repo["resolved"], resolved_change, "repo-root", legacy_repo["requested"]),
            ],
        )

    if explicit_state or legacy_repo:
        root = explicit_state or legacy_repo
        if is_linked_worktree(root["resolved"], git_worktrees(root["resolved"])):
            return unresolved_context(
                code_root=resolved_code_root,
                change_id=resolved_change,
                source="unresolved",
                unresolved_reason="linked-worktree-state-root",
                is_linked_worktree=True,
                candidates=[candidate(root["resolved"], resolved_change, root["source"], root["requested"])],
            )
        return resolved_context(
            state_root=root["resolved"],
            code_root=resolved_code_root,
            change_id=resolved_change,
            source=root["source"],
            candidates=[candidate(root["resolved"], resolved_change, root["source"], root["requested"])],
        )

    env_root = resolve_explicit_root(env_map.get("HARNESS_CHANGE_STATE_ROOT"), cwd_path, "environment")
    if env_root:
        if env_root["requested"] != env_root["resolved"]:
            return unresolved_context(
                code_root=resolved_code_root,
                change_id=resolved_change,
                source="unresolved",
                unresolved_reason="aliased-state-root",
                candidates=[candidate(env_root["resolved"], resolved_change, "environment", env_root["requested"])],
            )
        if is_linked_worktree(env_root["resolved"], git_worktrees(env_root["resolved"])):
            return unresolved_context(
                code_root=resolved_code_root,
                change_id=resolved_change,
                source="unresolved",
                unresolved_reason="linked-worktree-state-root",
                is_linked_worktree=True,
                candidates=[candidate(env_root["resolved"], resolved_change, "environment", env_root["requested"])],
            )
        return resolved_context(
            state_root=env_root["resolved"],
            code_root=resolved_code_root,
            change_id=resolved_change,
            source="environment",
            candidates=[candidate(env_root["resolved"], resolved_change, "environment", env_root["requested"])],
        )

    worktrees = git_worktrees(resolved_code_root)
    linked = is_linked_worktree(resolved_code_root, worktrees)
    candidates: list[dict] = []
    known_worktree_candidates = []

    if resolved_change and cwd_change and cwd_change["change_id"] == resolved_change:
        add_candidate(candidates, Path(cwd_change["state_root"]), resolved_change, "cwd-change", True)
    if resolved_change:
        add_candidate(candidates, resolved_code_root, resolved_change, "code-root", True)
    for worktree in worktrees:
        known_worktree_candidates.append(candidate(worktree, resolved_change, "git-worktree"))
        if resolved_change and (worktree / ".changes" / resolved_change).exists():
            add_candidate(
                candidates,
                worktree,
                resolved_change,
                "code-root" if worktree == resolved_code_root else "git-worktree",
                True,
            )

    if len(candidates) > 1:
        return unresolved_context(
            code_root=resolved_code_root,
            change_id=resolved_change,
            source="unresolved",
            unresolved_reason="ambiguous-state-root",
            is_linked_worktree=linked,
            candidates=candidates,
        )

    if len(candidates) == 1:
        only = candidates[0]
        only_root = Path(only["state_root"]).resolve()
        if only_root != resolved_code_root and only["source"] == "git-worktree":
            return unresolved_context(
                code_root=resolved_code_root,
                change_id=resolved_change,
                source="unresolved",
                unresolved_reason="other-worktree-state-root",
                is_linked_worktree=linked,
                candidates=candidates,
            )
        return resolved_context(
            state_root=only_root,
            code_root=resolved_code_root,
            change_id=resolved_change,
            source=only["source"],
            is_linked_worktree=linked,
            candidates=candidates,
        )

    if linked:
        return unresolved_context(
            code_root=resolved_code_root,
            change_id=resolved_change,
            source="unresolved",
            unresolved_reason="linked-worktree-unresolved",
            is_linked_worktree=linked,
            candidates=known_worktree_candidates,
        )

    return resolved_context(
        state_root=resolved_code_root,
        code_root=resolved_code_root,
        change_id=resolved_change,
        source="cwd-change" if cwd_change else "cwd",
        is_linked_worktree=linked,
        candidates=candidates,
    )


def error_text(context: dict) -> str:
    change = f" for change {context['change_id']}" if context.get("change_id") else ""
    reason = context.get("unresolved_reason")
    if reason == "conflicting-explicit-roots":
        state = next((item["state_root"] for item in context["candidates"] if item["source"] == "state-root"), "<missing>")
        repo = next((item["state_root"] for item in context["candidates"] if item["source"] == "repo-root"), "<missing>")
        return f"ERROR: conflicting state roots{change}: --state-root {state} differs from --repo-root {repo}\nGUIDE: retry with one canonical state root."
    if reason == "aliased-state-root":
        item = context["candidates"][0] if context.get("candidates") else {}
        return f"ERROR: state root alias{change}: {item.get('requested_root', '<missing>')} resolves to {item.get('state_root', '<missing>')}\nGUIDE: retry with the canonical state-root path."
    if reason == "linked-worktree-state-root":
        item = context["candidates"][0] if context.get("candidates") else {}
        return f"ERROR: state root{change} is a linked worktree: {item.get('state_root', '<missing>')}\nGUIDE: retry with the canonical shared state-root, not a linked worktree."
    if reason == "ambiguous-state-root":
        return "\n".join([f"ERROR: ambiguous state root{change}", *candidate_lines(context), "GUIDE: inspect candidates and retry with --state-root <canonical-state-root>."])
    if reason == "other-worktree-state-root":
        return "\n".join([f"ERROR: state root unresolved{change}; matching change workspace is in another worktree", *candidate_lines(context), "GUIDE: retry with explicit --state-root after confirming the canonical workspace."])
    return "\n".join([f"ERROR: state root unresolved{change}", *candidate_lines(context), "GUIDE: retry with --state-root <canonical-state-root>."])


def resolved_context(*, state_root: Path, code_root: Path, change_id: str | None, source: str, is_linked_worktree: bool = False, candidates=None) -> dict:
    return {
        "state_root": str(Path(state_root).resolve()),
        "code_root": str(Path(code_root).resolve()),
        "change_id": change_id,
        "source": source,
        "is_linked_worktree": is_linked_worktree,
        "candidates": candidates or [],
        "unresolved_reason": None,
    }


def unresolved_context(*, code_root: Path, change_id: str | None, source: str, unresolved_reason: str, is_linked_worktree: bool = False, candidates=None) -> dict:
    return {
        "state_root": None,
        "code_root": str(Path(code_root).resolve()),
        "change_id": change_id,
        "source": source,
        "is_linked_worktree": is_linked_worktree,
        "candidates": candidates or [],
        "unresolved_reason": unresolved_reason,
    }


def candidate(state_root: Path, change_id: str | None, source: str, requested_root: Path | None = None) -> dict:
    state_root = Path(state_root).resolve()
    change_dir = state_root / ".changes" / change_id if change_id else None
    result = {
        "state_root": str(state_root),
        "change_dir": str(change_dir) if change_dir else None,
        "source": source,
        "exists": change_dir.exists() if change_dir else state_root.exists(),
    }
    if requested_root is not None:
        result["requested_root"] = str(requested_root)
    return result


def add_candidate(candidates: list[dict], state_root: Path, change_id: str, source: str, require_exists: bool) -> None:
    entry = candidate(state_root, change_id, source)
    if require_exists and not entry["exists"]:
        return
    if any(Path(existing["state_root"]).resolve() == Path(entry["state_root"]).resolve() for existing in candidates):
        return
    candidates.append(entry)


def candidate_lines(context: dict) -> list[str]:
    return [
        f"CANDIDATE: {item['state_root']} source={item['source']} exists={'yes' if item['exists'] else 'no'}"
        for item in context.get("candidates", [])
    ]


def resolve_root(value: str | None, cwd: Path) -> Path | None:
    root = resolve_explicit_root(value, cwd, "explicit")
    return root["resolved"] if root else None


def resolve_explicit_root(value: str | None, cwd: Path, source: str) -> dict | None:
    if not value:
        return None
    path = Path(value)
    if not path.is_absolute():
        path = cwd / path
    requested = Path(os.path.abspath(path))
    return {"requested": requested, "resolved": realpath_or_resolve(requested), "source": source}


def resolve_code_root(value: str | None, cwd: Path) -> Path:
    explicit = resolve_root(value, cwd)
    if explicit:
        return explicit
    return git_root(cwd) or cwd


def git_root(cwd: Path) -> Path | None:
    result = subprocess.run(["git", "-C", str(cwd), "rev-parse", "--show-toplevel"], text=True, capture_output=True, check=False)
    if result.returncode != 0:
        return None
    return Path(result.stdout.strip()).resolve()


def git_worktrees(code_root: Path) -> list[Path]:
    result = subprocess.run(["git", "-C", str(code_root), "worktree", "list", "--porcelain"], text=True, capture_output=True, check=False)
    if result.returncode != 0:
        return []
    return [
        Path(line.removeprefix("worktree ").strip()).resolve()
        for line in result.stdout.splitlines()
        if line.startswith("worktree ")
    ]


def is_linked_worktree(code_root: Path, worktrees: list[Path]) -> bool:
    if len(worktrees) <= 1:
        return False
    current = realpath_or_resolve(code_root)
    return any(realpath_or_resolve(worktree) == current for worktree in worktrees[1:])


def find_cwd_change(cwd: Path) -> dict | None:
    parts = cwd.resolve().parts
    for index in range(len(parts) - 2, -1, -1):
        if parts[index] != ".changes" or not parts[index + 1]:
            continue
        state_root = Path(*parts[:index]).resolve() if index > 0 else Path("/").resolve()
        return {
            "state_root": str(state_root),
            "change_id": parts[index + 1],
            "change_dir": str(state_root / ".changes" / parts[index + 1]),
        }
    return None


def realpath_or_resolve(path: Path) -> Path:
    try:
        return path.resolve(strict=True)
    except OSError:
        return path.resolve()
